ChainBridge documentation (#24)
0xPolygon/polygon-sdk-docs: docs/chainbridge/setup/setup-erc20-transfer.md

---
id: setup-erc20-transfer
title: ERC20 Token Transfer
---

> So far, we've set up bridge to exchange assets/data between Polygon PoS and Polygon SDK chain. This section guides you a way how to set up ERC20 bridge and send tokens to another chain.

Reviewer: ```suggestion
So far, we've set up bridge to exchange assets/data between Polygon PoS and Polygon SDK chain. This section will guide you to set up an ERC20 bridge and send tokens between different blockchains.
```

Author: Fixed at 5756c760532fdfa7da42c47aad4dc893d1e4074e

## Step 1: Register resource ID

> Firstly, you will register a resource ID which associates resources in a cross-chain environment. A Resource ID is 32 bytes value which must be different from other resources but the same on every chain. The Resource IDs are arbitrary, but they may have the chain ID of the home chain in the last byte.

Reviewer: ```suggestion
Firstly, you will register a resource ID which associates resources in a cross-chain environment. A Resource ID is a 32 byte value which must be unique to the resource that we are transferring between these blockchains. The Resource IDs are arbitrary, but they may have the chain ID of the home chain in the last byte, as a convention (home chain referring to the network on which this resources originated from).
```

Author: Fixed at 5756c760532fdfa7da42c47aad4dc893d1e4074e

Author: Fixed at 5756c760532fdfa7da42c47aad4dc893d1e4074e

> To register resource ID, you can use `cb-sol-cli bridge register-resource` command. You will need to give the private key of `admin` account.

Reviewer: ```suggestion
To register resource ID, you can use the `cb-sol-cli bridge register-resource` command. You will need to give the private key of the `admin` account.
```

Author: Fixed at 5756c760532fdfa7da42c47aad4dc893d1e4074e

```bash
# For Polygon PoS chain
$ cb-sol-cli bridge register-resource \
  --url https://rpc-mumbai.matic.today \
  --privateKey [ADMIN_ACCOUNT_PRIVATE_KEY] \
  # Set Resource ID for ERC20
  --resourceId "0x000000000000000000000000000000c76ebe4a02bbc34786d860b355f5a5ce00" \
  --bridge "[BRIDGE_CONTRACT_ADDRESS]" \
  --handler "[ERC20_HANDLER_CONTRACT_ADDRESS]" \
  --targetContract "[ERC20_CONTRACT_ADDRESS]"

# For Polygon SDK chain
$ cb-sol-cli bridge register-resource \
  --url http://localhost:10002 \
  --privateKey [ADMIN_ACCOUNT_PRIVATE_KEY] \
  # Set Resource ID for ERC20
  --resourceId "0x000000000000000000000000000000c76ebe4a02bbc34786d860b355f5a5ce00" \
  --bridge "[BRIDGE_CONTRACT_ADDRESS]" \
  --handler "[ERC20_HANDLER_CONTRACT_ADDRESS]" \
  --targetContract "[ERC20_CONTRACT_ADDRESS]"
```

## (Optional) Make contracts mintable/burnable

> When you send ERC20 tokens between chains, tokens are processed by either way as follows:

Author: Fixed at 5756c760532fdfa7da42c47aad4dc893d1e4074e

> (1) Lock/Release mode

Reviewer: Provide some information on the Lock/Release mode, as to how the liquidity is provided on the "non-home" chain.

Author: Fixed at 5756c760532fdfa7da42c47aad4dc893d1e4074e

Author: @lazartravica In the "non-home" chain, the Handler contract has to have enough amount of Tokens beforehand if the contract is set to `Lock/Release` mode. The tokens won't be minted automatically and the Handler contract can transfer the same amount of the tokens in the Handler contract.

> **Source chain**: The tokens you're sending will be locked in ERC20 Handler Contract  

Reviewer: Avoid abbreviations (you're, you'll, I'll, etc.) in documentation

```suggestion
**Source chain**: The tokens you are sending will be locked in ERC20 Handler Contract  
```

Author: Fixed at 5756c760532fdfa7da42c47aad4dc893d1e4074e

> **Destination chain**: The same amount of tokens as you sent in the source chain would be unlocked and transferred from ERC20 Handler contract to the recipient account in the destination chain.

Reviewer: ```suggestion
**Destination chain**: The same amount of tokens as you sent and locked on the source chain will be unlocked and transferred from the ERC20 Handler contract to the recipient account on the destination chain.
```

Author: Fixed at 5756c760532fdfa7da42c47aad4dc893d1e4074e

(2) Burn/Mint mode

> **Source chain**: The tokens you're sending will be burned  

Reviewer: ```suggestion
**Source chain**: The tokens you are sending will be burned
```

Author: Fixed at 5756c760532fdfa7da42c47aad4dc893d1e4074e

> **Destination chain**: The same amount of tokens as you sent in the source chain would be minted in the destination chain and sent to the recipient account.

Reviewer: ```suggestion
**Destination chain**: The same amount of tokens as you sent and burned on the source chain will be minted on the destination chain and sent to the recipient account.
```

Author: Fixed at 5756c760532fdfa7da42c47aad4dc893d1e4074e

The default mode is **lock/release** mode. If you want to make the Tokens mintable/burnable, you need to call `adminSetBurnable` method. If you want to mint tokens on execution, you will need to grant `minter` role to ERC20 Handler contract.

```bash
# Let ERC20 contract burn on source chain and mint on destination chain
$ cb-sol-cli bridge set-burn \
  --url http://localhost:10002 \
  --privateKey [ADMIN_ACCOUNT_PRIVATE_KEY] \
  --bridge "[BRIDGE_CONTRACT_ADDRESS]" \
  --handler "[ERC20_HANDLER_CONTRACT_ADDRESS]" \
  --tokenContract "[ERC20_CONTRACT_ADDRESS]"

# Grant minter role to ERC20 Handler contract
$ cb-sol-cli erc20 add-minter \
  --url http://localhost:10002 \
  --privateKey [ADMIN_ACCOUNT_PRIVATE_KEY] \
  --erc20Address "[ERC20_CONTRACT_ADDRESS]" \
  --minter "[ERC20_HANDLER_CONTRACT_ADDRESS]"
```

## Step 2: Transfer ERC20 Token

We will send ERC20 Tokens from Polygon PoS chain to Polygon SDK chain.

> First, you will get tokens by minting. An account with `minter` role can mint new tokens. The account that has deployed ERC20 contract has `minter` role as default. For other accounts, you would need to run `cb-sol-cli erc20 add-minter` command to grant a minter role to the account before minting.

Reviewer: ```suggestion
First, you will get tokens by minting. An account with the `minter` role can mint new tokens. The account that has deployed the ERC20 contract has the `minter` role by default. To specify other accounts as members of the `minter` role, you need to run the `cb-sol-cli erc20 add-minter` command.
```

Reviewer: Specify also how often this needs to happen so that the users may know if this is something that can be done in advance to minimize gas costs

Author: Fixed at 5756c760532fdfa7da42c47aad4dc893d1e4074e

Author: @lazartravica Sorry. To be honest, I don't understand this well. In actual case, the user who wants to transfer tokens between chains won't mint tokens because the user should not be allowed to mint. Basically, the manager of the token contract will mint or the token contract schedules to mint.
